feat: Spring boot native image support

Christopher-Chianelli opened the pull request:
Spring boot does not seem to have a way to record POJOs to use them in generated files. I written a basic one that assumes all fields on POJOs have public setters and can serialize primitives, builtins, and collection types.

Christopher-Chianelli committed:
feat: Spring boot native image support

Spring boot does not seem to have a way to record POJOs to
use them in generated files. I written a basic one that
assumes all fields on POJOs have public setters and can
serialize primitives, builtins, and collection types.

Christopher-Chianelli commented:
Additionally, this uses reflection; future work would be to integrate GIZMO code.

Christopher-Chianelli committed:
fix: Add support for Duration code serialization when generating Spring Native image

Christopher-Chianelli committed:
chore: Move Pojo serialization logic to it own class and add tests for it

Christopher-Chianelli committed:
chore: Move all static initializer code to PojoInliner

Christopher-Chianelli committed:
ci: Install GraalVM native image in native workflow

This is to allow spring-integration-test to run
in native mode, as it requires a local GraalVM
native-image installation.

triceo commented on spring-integration/spring-boot-autoconfigure/src/main/java/ai/timefold/solver/spring/boot/autoconfigure/TimefoldBeanFactory.java:
```diff
@@ -0,0 +1,204 @@
+                final String noConstraintProviderErrorMsg = (scoreDirectorFactoryConfig != null)
+                        ? "Cannot provision a ConstraintVerifier because there is no ConstraintProvider class."
+                        : "Cannot provision a ConstraintVerifier because there is no PlanningSolution or PlanningEntity classes.";
+                return new ConstraintVerifier<>() {
```
Let's make this a proper class, inner if you prefer; reads better in stack traces and profiler results.

triceo commented on spring-integration/spring-boot-autoconfigure/src/main/java/ai/timefold/solver/spring/boot/autoconfigure/util/PojoInliner.java:
```diff
@@ -0,0 +1,419 @@
+        }
+        if (pojo instanceof Class<?> value) {
+            if (!Modifier.isPublic(value.getModifiers())) {
+                throw new IllegalArgumentException("Cannot serialize (" + value + ") because it is not a public class.");
```
```suggestion
                throw new IllegalArgumentException("Cannot serialize (%s) because it is not a public class."
                        .formatted(value));
```

Let's consistently apply this convention in new code.
It makes for messages that are easier to read, especially as the number of arguments grows.

triceo commented on spring-integration/spring-boot-integration-test/src/main/java/ai/timefold/solver/spring/boot/it/TimefoldController.java:
```diff
@@ -0,0 +1,25 @@
+
+@RestController
+@RequestMapping("/integration-test")
+public class TimefoldController {
```
Timefold is a company; we have Solver, we have Orbit. Let's start using `TimefoldSolver` consistently, because we're dealing with the solver here, not with the company.

triceo reviewed (changes requested):
Major comments:

- Serialization _must not_ be our business. If we don't know how to make it happen, we need to ask the Spring guys. I can not accept responsibility for maintaining our own serialization framework.
- Please use Timefold Solver consistently. Where you're speaking of `Timefold`, you probably should have been speaking of `TimefoldSolver`.
- Please include a section in our documentation for Spring on how to build a native image. Especially mention all the quirks of our implementation which may not be immediately obvious, such as any requirements to make anything public.
- Please make sure that the Spring quickstart supports native, and that it is tested as part of the standard PR workflow on the Quickstarts repo.

Christopher-Chianelli commented on spring-integration/spring-boot-autoconfigure/src/main/java/ai/timefold/solver/spring/boot/autoconfigure/util/PojoInliner.java:
```diff
@@ -0,0 +1,419 @@
+import org.springframework.javapoet.CodeBlock;
+import org.springframework.javapoet.TypeSpec;
+
+public final class PojoInliner {
```
Agreed, but sadly using XML serialization causes GraalVM to complain about a JAR file being in the native heap, even when the XML serialization is only used for writing the String (outside the generated class with no parsing!)

triceo replied:
I see your point. Still, this doesn't change my mind - I am not yet convinced we have exhausted all possibilities. 

Please go to the Spring guys, explain what you need to do, and ask them what they think would be the proper solution. If they say that this is the best we can possibly do, then I have no choice but to accept it.

I suggest you ask Geoffrey, he can introduce you to some people.

Christopher-Chianelli replied:
Well, apparently registering too many things for reflection is a bad thing. GraalVM did not like when `java.lang.Class` was registered for reflection, since it made it think its cachedHashCode field was reachable. Not registering java.lang.Class and java.lang.ClassLoader does make it work (although JAXB accesses itself using reflection, so I opted for Jackson instead)

Christopher-Chianelli commented on spring-integration/spring-boot-autoconfigure/src/main/java/ai/timefold/solver/spring/boot/autoconfigure/TimefoldAotContribution.java:
```diff
@@ -0,0 +1,200 @@
+import org.springframework.javapoet.CodeBlock;
+import org.springframework.javapoet.MethodSpec;
+
+public class TimefoldAotContribution implements BeanFactoryInitializationAotContribution {
```
It ends in `AotContribution` because it implements `BeanFactoryInitializationAotContribution`; will add `Solver`

Christopher-Chianelli committed:
chore: Rename Timefold -> TimefoldSolver where it makes sense

Also changed the UnsupportedConstraintVerifier class from
an anonymous class to an inner class.

Christopher-Chianelli committed:
chore: Rename Timefold -> TimefoldSolver in spring-boot-integration-test

Christopher-Chianelli committed:
chore: Use ObjectMapper instead of PojoInliner to transfer SolverConfig to generated method in spring

Finding out why ObjectMapper wasn't before was trickly. The short of it
is that java.lang.Class was registered for reflection, which made
GraalVM think its cachedHashCode field was reachable in an initializer,
which is not allowed. So we needed to make registerTypeRecursively
ignore java.lang.Class and java.lang.ClassLoader. Then ObjectMapper is
happy. I would use SolverConfigIO/JAXB, but JAXB accesses itself using
reflection, so I would need to register a bunch more (possible unstable)
classes for reflections to use it.

Christopher-Chianelli committed:
chore: Use Spring idioms, add XML reflection config files

Christopher-Chianelli committed:
chore: Reduce the number of items in reflect-config.json

The no-args default constructor is not included in "queryAllDeclaredMethods".

Christopher-Chianelli committed:
chore: Simplify reflect-config.json and add a large spring native XML test

triceo commented:
Some outstanding things before we can merge:

- [ ] I changed our Sonar settings, hopefully removing many of the things we don't want to worry about. Please take a look at what remains and resolve it before merging.
- [ ] See Geoffrey's and my comments on Slack.
- [ ] It shows me that the branch has conflicts.
- [ ] The quickstarts PR is not yet ready for review.

Christopher-Chianelli committed:
chore: Address SonarCloud issues, change score classes -> score calculator classes in Spring error message
